quil-rs/src/program/mod.rs

```diff
@@ -348,6 +356,108 @@ impl Program {
+    ///
+    /// See [`resolve_placeholders_with_custom_resolvers`](Self::resolve_placeholders_with_custom_resolvers),
+    /// [`default_label_resolver`](Self::default_label_resolver),
+    /// and [`default_qubit_resolver`](Self::default_qubit_resolver) for more information.
+    pub fn resolve_placeholders(&mut self) {
+        self.resolve_placeholders_with_custom_resolvers(
+            self.default_label_resolver(),
+            self.default_qubit_resolver(),
+        )
+    }
+
+    /// Resolve [`LabelPlaceholder`]s and [`QubitPlaceholder`]s within the program such that the resolved values
+    /// will remain unique to that placeholder within the scope of the program.
+    ///
+    /// The provided `label_resolver` and `qubit_resolver`, will be used to resolve those values respectively.
+    /// If your placeholder returns `None` for a particular placeholder, it will not be replaced but will be left as a placeholder.
+    ///
+    /// If you wish to provide a resolver for either labels or qubits, but want to rely on the
+    /// default behavior for the other, considering using either
+    /// [`default_qubit_resolver`](Self::default_qubit_resolver) or [`default_label_resolver`](Self::default_label_resolver).
+    #[allow(clippy::type_complexity)]
+    pub fn resolve_placeholders_with_custom_resolvers(
+        &mut self,
+        label_resolver: Box<dyn Fn(&LabelPlaceholder) -> Option<String>>,
+        qubit_resolver: Box<dyn Fn(&QubitPlaceholder) -> Option<u64>>,
+    ) {
+        for instruction in &mut self.instructions {
+            instruction.resolve_placeholders(&label_resolver, &qubit_resolver);
+        }
+    }
+
+    /// The default label resolver will resolve each [`LabelPlaceholder`] in the program to a unique label
+    /// by applying an auto-incrementing suffix to the base label.
+    pub fn default_label_resolver(&self) -> Box<dyn Fn(&LabelPlaceholder) -> Option<String>> {
+        let mut fixed_labels = HashSet::new();
+        let mut label_placeholders = IndexSet::new();
+        for label in self.get_labels() {
+            match label {
+                Label::Fixed(fixed) => {
+                    fixed_labels.insert(fixed.clone());
+                }
+                Label::Placeholder(placeholder) => {
+                    label_placeholders.insert(placeholder.clone());
+                }
+            }
+        }
+
+        let label_resolutions: HashMap<LabelPlaceholder, String> = label_placeholders
+            .into_iter()
+            .map(|p| {
+                let base_label = p.as_inner();
+                let mut next_suffix = 0;
+
+                loop {
+                    let next_label = format!("{base_label}_{next_suffix}");
+
+                    if !fixed_labels.contains(&next_label) {
+                        fixed_labels.insert(next_label.clone());
+                        break (p, next_label);
```

Comment:
I don't think there is. IMO `return` would be clearer here.

Also, it might be clearer to do something like

```rust
for next_suffix in 0.. {
    let next_label = ...;
    if !fixed_labels.contains(&next_label) {
        // ...
    }
}
```

Reply:
Took a different approach that I think expresses the sentinel condition the most clearly.